File: content/snippets/_harness/fake-llm.mjs
/**
 * Local double for a general-purpose LLM API. JavaScript counterpart of
 * _harness/fake_llm.py. See that file for why this exists.
 */

export class FakeLLMError extends Error {}

export class FakeLLM {
  /**
   * @param {object} options
   * @param {string|object} [options.response] what `complete` should return
   * @param {number} [options.failTimes] how many calls should fail first
   * @param {string} [options.model]
   */
  constructor({ response = '', failTimes = 0, model = 'fake-model' } = {}) {
    this.response = response;
    this.model = model;
    this.requests = [];
    this._failTimes = failTimes;
  }

  get lastRequest() {
    if (this.requests.length === 0) {
      throw new Error('the snippet never called the client');
    }
    return this.requests.at(-1);
  }

  get callCount() {
    return this.requests.length;
  }

  _record(request) {
    this.requests.push(request);
    if (this._failTimes > 0) {
      this._failTimes -= 1;
      throw new FakeLLMError('simulated provider failure');
    }
  }

  async complete(request) {
    this._record(request);
    return typeof this.response === 'string' ? this.response : JSON.stringify(this.response);
  }

  async completeJson(request) {
    this._record(request);
    return typeof this.response === 'string' ? JSON.parse(this.response) : this.response;
  }

  /**
   * Deterministic pseudo-embeddings. Same text gives the same vector, and
   * texts sharing words land closer together.
   */
  async embed(texts, options = {}) {
    this._record({ texts, ...options });
    return texts.map((t) => bagOfWordsVector(t));
  }
}

/**
 * FNV-1a, 32 bits. Must agree exactly with its Python counterpart, so that a
 * snippet's two language versions score the same inputs identically.
 *
 * Math.imul is what makes that true: a plain `*` on numbers this large loses
 * precision past 2^53 and silently drifts from Python's exact integers.
 */
export function hashWord(word) {
  let h = 2166136261;
  for (const char of word) {
    h = Math.imul(h ^ char.codePointAt(0), 16777619) >>> 0;
  }
  return h;
}

function bagOfWordsVector(text, dimensions = 16) {
  const vector = new Array(dimensions).fill(0);
  for (const word of text.toLowerCase().split(/\s+/).filter(Boolean)) {
    vector[hashWord(word) % dimensions] += 1;
  }
  const norm = Math.hypot(...vector);
  return norm ? vector.map((v) => v / norm) : vector;
}

File: content/snippets/_harness/fake_llm.py
"""
Local double for a general-purpose LLM API.

Snippets on rung N3 call a real provider. Their tests cannot: that would cost
money, need a key, and break the build whenever the provider is down. So the
test injects this double instead.

What the resulting test actually proves: the request is built correctly, the
response is decoded correctly, and error paths are handled. That is where the
bugs in this kind of code live.

What it does not prove: that the model answers well. Snippets tested this way
are declared `verification: stubbed` on their entry, and the page says so.
"""

from __future__ import annotations

import json
from typing import Any


class FakeLLMError(Exception):
    """Raised by the double when it is told to simulate a provider failure."""


class FakeLLM:
    """
    Records what it is asked and returns what it was told to return.

    >>> client = FakeLLM(response="hello")
    >>> client.complete(prompt="say hi")
    'hello'
    >>> client.last_request["prompt"]
    'say hi'
    """

    def __init__(
        self,
        response: str | dict[str, Any] = "",
        *,
        fail_times: int = 0,
        model: str = "fake-model",
    ) -> None:
        self.response = response
        self.model = model
        self.requests: list[dict[str, Any]] = []
        self._fail_times = fail_times

    @property
    def last_request(self) -> dict[str, Any]:
        if not self.requests:
            raise AssertionError("the snippet never called the client")
        return self.requests[-1]

    @property
    def call_count(self) -> int:
        return len(self.requests)

    def _record(self, **kwargs: Any) -> None:
        self.requests.append(kwargs)
        if self._fail_times > 0:
            self._fail_times -= 1
            raise FakeLLMError("simulated provider failure")

    def complete(self, **kwargs: Any) -> str:
        """Text completion. Mirrors the shape of a chat completion call."""
        self._record(**kwargs)
        if isinstance(self.response, dict):
            return json.dumps(self.response)
        return self.response

    def complete_json(self, **kwargs: Any) -> Any:
        """Structured output. Returns the response already decoded."""
        self._record(**kwargs)
        if isinstance(self.response, dict):
            return self.response
        return json.loads(self.response)

    def embed(self, texts: list[str], **kwargs: Any) -> list[list[float]]:
        """
        Deterministic pseudo-embeddings.

        Same text gives the same vector, and texts sharing words land closer
        together, which is enough to exercise a nearest-neighbour code path.
        """
        self._record(texts=texts, **kwargs)
        return [_bag_of_words_vector(t) for t in texts]


def _bag_of_words_vector(text: str, dimensions: int = 16) -> list[float]:
    vector = [0.0] * dimensions
    for word in text.lower().split():
        vector[hash_word(word) % dimensions] += 1.0
    norm = sum(v * v for v in vector) ** 0.5
    return [v / norm for v in vector] if norm else vector


def hash_word(word: str) -> int:
    """Stable across runs, unlike the built-in hash of a string."""
    h = 2166136261
    for char in word:
        h = ((h ^ ord(char)) * 16777619) & 0xFFFFFFFF
    return h
